WooExpress: Add WaitForAtomic step

markbiek opened the pull request:
#### Proposed Changes

Similar to how we handle the tailored ecommerce flow, this adds the `WaitForAtomic` step to the WooExpress flow to ensure that we don't redirect to the dashboard until the Atomic transfer is complete.

#### Testing Instructions

1. Apply this branch and visit http://calypso.localhost:3000/setup/wooexpress/.
2. You should go through the following steps
  a. `siteCreationStep`
  b. `processing`
  c. `assignTrialPlan`
  d. `waitForAtomic`
  e. `processing`
3. Once that's done, you should be redirected to the dashboard and the site should be fully transferred to Atomic.

#### Pre-merge Checklist

<!--
Complete applicable items on this checklist **before** merging into trunk. Inapplicable items can be left unchecked.

Both the PR author and reviewer are responsible for ensuring the checklist is completed.
-->

- [x] Has the general commit checklist been followed? (PCYsg-hS-p2)
- [NA] [Have you written new tests](https://wpcalypso.wordpress.com/devdocs/docs/testing/index.md) for your changes?
- [NA] Have you tested the feature in Simple (P9HQHe-k8-p2), Atomic (P9HQHe-jW-p2), and self-hosted Jetpack sites (PCYsg-g6b-p2)?
- [x] Have you checked for TypeScript, React or other console errors?
- [NA] Have you used memoizing on expensive computations? More info in [Memoizing with create-selector](https://github.com/Automattic/wp-calypso/blob/trunk/packages/state-utils/src/create-selector/README.md) and [Using memoizing selectors](https://react-redux.js.org/api/hooks#using-memoizing-selectors) and [Our Approach to Data](https://github.com/Automattic/wp-calypso/blob/trunk/docs/our-approach-to-data.md)
- [NA] Have we added the "[Status] String Freeze" label as soon as any new strings were ready for translation (p4TIVU-5Jq-p2)?
- [NA] For changes affecting Jetpack: Have we added the "[Status] Needs Privacy Updates" label if this pull request changes what data or activity we track or use (p4TIVU-ajp-p2)?
<!--
Link a related issue to this PR. If the PR does not immediately resolve the issue,
for example, it requires a separate deployment to production, avoid
using the "fixes" keyword and instead attach the [Status] Fix Inbound label to
the linked issue.
-->

Closes #71868

markbiek committed:
Add WaitForAtomic step to the WooExpress flow

markbiek commented:
Wanted to note that the functionality for this is working but there's some inconsistency with the progress bar at the top of each step. I'm working on a fix for that now.

sixhours commented on client/landing/stepper/declarative-flow/internals/steps-repository/wait-for-atomic/index.tsx:
```diff
@@ -32,13 +32,17 @@ export const transferStates = {
 
-	const siteId = site?.ID;
+	let siteId = site?.ID as number;
+	// In some cases, we get the siteId from the navigation extra data rather than the SITE_STORE.
```
Why (out of curiosity)?

markbiek replied:
Because the WooExpress flow doesn't require any user interaction, it doesn't use the data stores in the same way as the other flows. So instead of a user initiating the next step and it populating the `SITE_STORE`, the WooExpress `siteCreationStep` passes the newly created site information on to the next step via the extra-data functionality we have.

sixhours commented on client/landing/stepper/declarative-flow/trial-wooexpress-flow.ts:
```diff
@@ -76,10 +78,9 @@ const wooexpress: Flow = {
 		const storeType = useSelect( ( select ) => select( ONBOARD_STORE ).getStoreType() );
 
 		const flowProgress = useSiteSetupFlowProgress( currentStep, intent, storeType );
```
I'm wondering why we need the `flowProgress` piece for this flow (that's the large full-width progress bar at the top of the screen, right?), since it's not a flow with steps that the user can interact with or control. The current behavior is a bit wonky:

https://user-images.githubusercontent.com/2124984/211892054-f0dcd45e-dfd8-4dfe-a20e-76df993beb63.mov

Progress bar starts complete, goes back to the beginning, goes back to complete.

We also show a second progress bar during one of the `processing` steps, and a different spinner effect with the pulsing dots under the text. The overall experience is confusing :D

markbiek replied:
Agreed that the differing spinners is confusing but that's a known issue and there's been discussion about how/when to tackle it. We have the same problem with the tailored ecommerce flow.

As far as the progress bar itself, I can see if I can hide it entirely.

sixhours reviewed (commented):
This is working functionally! I just have a couple questions about the state of the progress bars throughout this flow.

markbiek committed:
Hide progress bar on wooexpress

Not needed since the user doesn't have any control over how they move through the steps

sixhours reviewed (approved):
Still working for me!
